src/categorical_algebra/DPO.jl

```
@@ -0,0 +1,177 @@
+    [ g_inv[lm[c](x)] for x in parts(I,c) ]
+  end)
+
+  k = ACSetTransformation(k_components, I, K)
+  g = ACSetTransformation(g_components, K, G)
+  return k => g
+end
+
+"""
+Apply a rewrite rule (given as a span, L<-I->R) to a ACSet
+using a match morphism `m` which indicates where to apply
+the rewrite.
+"""
+function rewrite_match(L::ACSetTransformation{CD, AD},
+                       R::ACSetTransformation{CD, AD},
+                       m::ACSetTransformation{CD, AD}
+                      )::AbstractACSet{CD, AD} where {CD, AD}
+    @assert dom(L) == dom(R)
+    @assert codom(L) == dom(m)
+    (k, _) = pushout_complement(L, m)
+    l1, _ = pushout(R, k)
+    return codom(l1)
+end
+
+"""
+Apply a rewrite rule (given as a span, L<-I->R) to a ACSet,
+where a match morphism is found automatically. If multiple
+matches are found, a particular one can be selected using
+`m_index`.
+"""
+function rewrite(L::ACSetTransformation{CD, AD},
+                 R::ACSetTransformation{CD, AD},
+                 G::AbstractACSet{CD, AD},
+                 monic::Bool=false, m_index::Int=1
+                )::Union{Nothing, AbstractACSet} where {CD, AD}
+  ms = filter(m->valid_dpo(L, m), homomorphisms(codom(L), G, monic=monic))
+  if 0 < m_index <= length(ms)
+    return rewrite_match(L, R, ms[m_index])
+  else
+    return nothing
+  end
+end
+
+
+"""
+Condition for existence of a pushout complement
+"""
+function valid_dpo(L::ACSetTransformation, m::ACSetTransformation; fail::Bool=false)::Bool
+  return id_condition(L, m;fail=fail) && dangling_condition(L, m;fail=fail)
+end
+
+"""
+Does not map both a deleted item and a preserved item in L to the same item in G, or two distinct deleted items to the same.
+(Trivially satisfied if match is mono)
+"""
+function id_condition(L::ACSetTransformation{CD, AD},
+                      m::ACSetTransformation{CD, AD};
+                      fail::Bool=false)::Bool where {CD, AD}
```

More useful than a debug switch that throws an error would be a switch to returns one or more pairs of offending elements when the ID conditions fails and `nothing` otherwise. Thus, you could actually use this information programmatically if you wanted to.

Also, when you *do* want to raise an error like this, use `error(...)` or `throw ...` rather than `@assert false`.